crates/metal-llvm/src/ty.rs

```diff
@@ -0,0 +1,56 @@
+use llvm_sys::core::{LLVMArrayType2, LLVMFunctionType, LLVMStructTypeInContext};
+use metal_mir::types::{function::FunctionSignature, Type};
+
+use super::{get_types, CodeGenType};
+
+impl CodeGenType for FunctionSignature {
+    fn codegen_type(
+        &self,
+        llvm: &crate::LLVMRefs,
+        module: &metal_mir::parcel::Module,
+    ) -> llvm_sys::prelude::LLVMTypeRef {
+        unsafe {
+            let len = self.arguments.len();
+            LLVMFunctionType(
+                self.return_type.codegen_type(llvm, module),
+                get_types(llvm, module, &self.arguments).as_mut_ptr(),
+                len.try_into().unwrap(),
+                0,
+            )
+        }
+    }
+}
+
+impl CodeGenType for Type {
+    fn codegen_type(
+        &self,
+        llvm: &crate::LLVMRefs,
+        module: &metal_mir::parcel::Module,
+    ) -> llvm_sys::prelude::LLVMTypeRef {
+        match self {
+            Self::Primitive(p) => p.codegen_type(llvm, module),
+            Self::Composite(c) => unsafe {
+                // Has to be cloned since we can't Unpack a referenced Box.
+                // Essentially: `**lit` doesn't work, since Copy isn't implemented
+                // for MIR types.
+                // Context: https://github.com/metallang/metal/pull/51#discussion_r1838137310
+                // https://github.com/metallang/metal/pull/51#discussion_r1838160041
+                match *c.clone() {
```

Comment:
This doesn't have to be cloned as an owned value is not required in any of the `match`'s arms.

```suggestion
                match &c {
```